### ISSUE TITLE
Spack Bootstrap to `/` (Docker)

### ISSUE THREAD

**Issue by ax3l** [LLNL/spack#4062]
I was wondering, while provisioning a docker image, if `spack bootstrap /` is actually something being considered as good practice / possible workflow. I could imagine so.

Because currently it fails complaining that there already files in `/` instead of adding its own :-)

Also it looks like bootstrap just downloads the source yet again which makes the quick-start quite long (git clone, ~~export path,~~ bootstrap, export path).
Do you plan to provide a simple all-in-one bootstrap script (e.g. as in [yt](http://yt-project.org/#getyt) or [anaconda](https://www.continuum.io/downloads)) to get users kick-started?

**Comment by tgamblin** [comment]
I like this idea.  Want to submit a PR and some docs with recommendations for using spack in containers?

**Comment by adamjstewart** [comment]
What is `spack bootstrap` for?

**Comment by tgamblin** [comment]
For sticking spack in a system prefix.  It's old and hasn't gotten a lot of love since the early incarnations of spack, but spack has always been structured to coexist with other things inside a prefix.

**Comment by adamjstewart** [comment]
Ah, I see. I didn't realize that you can't `git clone` into an existing directory. By the way, #3057 plans on renaming `spack bootstrap` to `spack clone` and creating a new `spack bootstrap` command to install Spack's run-time dependencies like environment modules.

**Comment by ax3l** [comment]
Actually, `spack bootstrap` also just works when executed on a `.git` repo dir, my test failed when doing it on a `.tar.gz` download from a tag :)

**Comment by ax3l** [comment]
> I like this idea. Want to submit a PR and some docs with recommendations for using spack in containers?

Dang, I planned to hide the container workflow task in a follow-up issue and then fade into darkness ;-)

Actually I have some questions on that: container provisioning aims to be slick and small (in terms of the size/diff of "docker layers"): what can I remove after a successful `spack install A && spack install B && spack install C` ~~orgy~~ install chain?

For example `apt-get update/install` commands are usually followed by `rm -rf /var/lib/apt/lists/*` and in the spack case I would like to guarantee all
- downloaded sources
- tmp build files
- caches
- ... (more?)

are removed after install to keep the diff small.

**Comment by adamjstewart** [comment]
> what can I remove

You can uninstall any build-only dependencies. I would also recommend running:
```
$ spack purge --all
```
See:
```
$ spack purge --help
usage: spack purge [-h] [-s] [-d] [-m] [-a]

optional arguments:
  -h, --help        show this help message and exit
  -s, --stage       remove all temporary build stages (default)
  -d, --downloads   remove cached downloads
  -m, --misc-cache  remove long-lived caches, like the virtual package index
  -a, --all         remove all of the above
```

**Comment by ax3l** [comment]
@tgamblin is it ok if I mark the two `spack bootstrap` tasks in this issue as bugs instead of features? ^^

**Comment by tgamblin** [comment]
sure!

**Comment by ax3l** [comment]
note: `spack purge --all` is now `spack clean -a`

**Comment by tgamblin** [comment]
@ax3l: is this still an issue?  Note that it's also called `spack clone` now.

**Comment by ax3l** [comment]
@tgamblin yes, the problem is still there as of `develop` @ 3e7680e24ad95e8fc30acb703c664e397375d135

(starting from a tarball download)
```bash
$SPACK_ROOT/bin/spack clone /usr
fatal: Not a git repository (or any of the parent directories): .git
==> Warning: No branch found; using default branch: develop
==> Warning: No git repository found; using default upstream URL: https://github.com/llnl/spack
==> Error: There are already files there! Delete these files before boostrapping spack.
  local
  etc
  include
  bin
  share
  lib
  sbin
  games
  src
```

**Comment by alalazo** [comment]
@ax3l I suppose we can close this since we have Docker images on Dockerhub already, right?

**Comment by ax3l** [comment]
Not sure, still think that bootstrapping into something like `/usr/local` would generally be cool.

**Comment by alalazo** [comment]
Fair enough, thanks @ax3l

**Comment by alalazo** [comment]
`spack clone` is no more https://github.com/spack/spack/pull/50883

**Comment by alalazo** [comment]
I'd be tempted to close this, and reopen a new issue in case something is still wanted from the list in the discussion. Let me know.
